src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -14,35 +14,289 @@ namespace Microsoft.CodeAnalysis
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReferencesMap,
+            ProjectId projectId,
+            IReadOnlyList<ProjectId> referencedProjectIds)
+        {
+            if (existingReferencesMap.TryGetValue(projectId, out var existingReferences))
+            {
+                return existingReferencesMap.SetItem(projectId, existingReferences.Union(referencedProjectIds));
+            }
+            else
+            {
+                return existingReferencesMap.SetItem(projectId, referencedProjectIds.ToImmutableHashSet());
+            }
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_lazyReverseReferencesMap"/> for the addition of additional project references. Must be called on a non-null
+        /// map.
+        /// </summary>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReverseReferencesMapForAdditionalProjectReferences(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReverseReferencesMap,
+            ProjectId projectId,
+            IReadOnlyList<ProjectId> referencedProjectIds)
+        {
+            var builder = existingReverseReferencesMap.ToBuilder();
+
+            foreach (var referencedProject in referencedProjectIds)
+            {
+                if (builder.TryGetValue(referencedProject, out var reverseReferences))
+                {
+                    builder[referencedProject] = reverseReferences.Add(projectId);
+                }
+                else
+                {
+                    builder[referencedProject] = ImmutableHashSet.Create(projectId);
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_transitiveReferencesMap"/> for the addition of additional project references. 
+        /// </summary>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewTransitiveReferencesMapForAdditionalProjectReferences(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingTransitiveReferencesMap,
+            ProjectId projectId,
+            IReadOnlyList<ProjectId> referencedProjectIds)
+        {
+            // To update our forward transitive map, we need to add referencedProjectIds (and their transitive dependencies) to the transitive references
+            // of projects. First, let's just compute the new set of transitive references. It's possible while doing so we'll discover that we don't
+            // know the transitive project references for one of our new references. In that case, we'll use null as a sentinel to mean "we don't know" and
+            // we propogate the not-knowingness. But let's not worry about that yet. First, let's just get the new transitive reference set.
+            var newTransitiveReferences = new HashSet<ProjectId>(referencedProjectIds);
```

Comment:
>new HashSet<ProjectId> [](start = 42, length = 22)

Can we use pooled HashSet here?

Reply:
Possibly. We'll let future measuring guide.